### PR TITLE
EDX-4239 Add partial go-mod-bootstrap code

### PR DESCRIPTION
**Pull request by jinlinGuan**

### CODE REVIEW

**Review by cloudxxx8 — CHANGES_REQUESTED** [review]

File: pkg/di/container.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	if service.instance == nil {
+		service.instance = service.constructor(c.get)
+		c.serviceMap[serviceName] = service
+	}
+	return service.instance
+}
+
+// Get wraps get to make it thread-safe.
+func (c *Container) Get(serviceName string) interface{} {
+	c.mutex.Lock()
```
Comment (cloudxxx8) [inline pkg/di/container.go:134]: Get should be `Rlock()`

**Review by cloudxxx8 — CHANGES_REQUESTED** [review]

File: pkg/di/type.go
Comment (cloudxxx8) [inline pkg/di/type.go:1]: rename to helper.go, because it doesn't contain any type.

**Review by cloudxxx8 — CHANGES_REQUESTED** [review]

File: pkg/secrets/types/config.go
Comment (cloudxxx8) [inline pkg/secrets/types/config.go:1]: remove this file and package
we don't use it for now

File: pkg/bootstrap/config/types.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	// TokenFile provides a location to a token file.
+	TokenFile string
+	// SecretsFile is optional Path to JSON file containing secrets to seed into service's SecretStore
+	SecretsFile string
+	// DisableScrubSecretsFile specifies to not scrub secrets file after importing. Service will fail start-up if
+	// not disabled and file can not be written.
+	DisableScrubSecretsFile bool
+
+	// RuntimeTokenProvider is optional if not using delayed start from spiffe-token provider
+	RuntimeTokenProvider types.RuntimeTokenProviderInfo
```
Comment (cloudxxx8) [inline pkg/bootstrap/config/types.go:71]: remove RuntimeTokenProvider

**Review by cloudxxx8 — CHANGES_REQUESTED** [review]

File: pkg/bootstrap/flags/flags.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+func (d *Default) Help() {
+	d.helpCallback()
+}
+
+// commonHelpCallback displays the help usage message and exits
+func (d *Default) helpCallback() {
+	fmt.Printf(
+		"Usage: %s [options]\n"+
+			"Server Options:\n"+
+			"    -cf, --configFile <name>        Indicates name of the local configuration file. Defaults to configuration.toml\n"+
```
[the remark this answers is not in the document]
Comment (jinlinGuan) [reply pkg/bootstrap/flags/flags.go]: updated

File: pkg/bootstrap/bootstrap.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	dic *di.Container,
+	useSecretProvider bool,
+	handlers []interfaces.BootstrapHandler) (*sync.WaitGroup, Deferred, bool) {
+
+	var err error
+	var wg sync.WaitGroup
+	deferred := func() {}
+
+	// Check if service provided an initial Logger to use. If not create one and add it to the DIC.
+	lc := container.LoggingClientFrom(dic.Get)
```
[the remark this answers is not in the document]
Comment (jinlinGuan) [reply pkg/bootstrap/bootstrap.go]: renamed

File: pkg/logger/logger.go [hunk 5]
```diff
@@ -10,11 +10,47 @@ import (
...
+const (
+	TraceLog = "TRACE"
+	DebugLog = "DEBUG"
+	InfoLog  = "INFO"
+	WarnLog  = "WARN"
+	ErrorLog = "ERROR"
+)
+
+// LoggingClient defines the interface for logging operations.
+type LoggingClient interface {
```
[the remark this answers is not in the document]
Comment (jinlinGuan) [reply pkg/logger/logger.go]: renamed

File: pkg/logger/logger.go [hunk 6]
```diff
@@ -10,11 +10,47 @@ import (
...
+	Errorf(msg string, args ...interface{})
+	// Infof logs a formatted message at the INFO severity level
+	Infof(msg string, args ...interface{})
+	// Tracef logs a formatted message at the TRACE severity level
+	Tracef(msg string, args ...interface{})
+	// Warnf logs a formatted message at the WARN severity level
+	Warnf(msg string, args ...interface{})
+}
+
 type Logger struct {
```
[the remark this answers is not in the document]
Comment (jinlinGuan) [reply pkg/logger/logger.go]: renamed

File: pkg/bootstrap/config/types.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+func NewSecretStoreInfo(serviceKey string) SecretStoreInfo {
```
[the remark this answers is not in the document]
Comment (jinlinGuan) [reply pkg/bootstrap/config/types.go]: removed

File: pkg/bootstrap/configprocessor/processor.go [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+		err = utils.ConvertFromMap(data, config)
+	case jsonExt:
+		err = json.Unmarshal(contents, config)
+	}
+
+	if err != nil {
+		return fmt.Errorf("failed to unmarshall configuration file %s: %s", filePath, err.Error())
+	}
+
+	cp.lc.Info(fmt.Sprintf("Loaded %s configuration from %s", configType, filePath))
```
[the remark this answers is not in the document]
Comment (jinlinGuan) [reply pkg/bootstrap/configprocessor/processor.go]: updated

File: pkg/bootstrap/configprocessor/processor.go [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+	case yamlExt, ymlExt:
+		data := make(map[string]any)
+		err = yaml.Unmarshal(contents, &data)
+		if err != nil {
+			return fmt.Errorf("failed to unmarshall configuration file %s: %s", filePath, err.Error())
+		}
+		err = utils.ConvertFromMap(data, config)
+	case jsonExt:
+		err = json.Unmarshal(contents, config)
+	}
```
[the remark this answers is not in the document]
Comment (jinlinGuan) [reply pkg/bootstrap/configprocessor/processor.go:126]: added

File: pkg/bootstrap/configprocessor/processor.go [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+		return fmt.Errorf("failed to unmarshall configuration file %s: %s", filePath, err.Error())
+	}
+
+	cp.lc.Info(fmt.Sprintf("Loaded %s configuration from %s", configType, filePath))
+
+	return nil
+}
+
+// getSecretNamesChanged returns a slice of secretNames that have changed secrets or are new.
+func getSecretNamesChanged(prevVals config.InsecureSecrets, curVals config.InsecureSecrets) []string {
```
[the remark this answers is not in the document]
Comment (jinlinGuan) [reply pkg/bootstrap/configprocessor/processor.go]: removed

File: pkg/di/container.go (hunk 1 above)
Comment (jinlinGuan) [reply pkg/di/container.go]: updated

File: pkg/di/type.go
Comment (jinlinGuan) [reply pkg/di/type.go:1]: renamed

File: pkg/secrets/types/config.go
Comment (jinlinGuan) [reply pkg/secrets/types/config.go:1]: removed

File: pkg/bootstrap/config/types.go (hunk 2 above)
Comment (jinlinGuan) [reply pkg/bootstrap/config/types.go]: removed

File: pkg/bootstrap/container/logger.go [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+import (
+	"github.com/IOTechSystems/go-mod-edge-utils/pkg/di"
+	"github.com/IOTechSystems/go-mod-edge-utils/pkg/logger"
+)
+
+// LoggingClientInterfaceName contains the name of the logger.LoggingClient implementation in the DIC.
+var LoggingClientInterfaceName = di.TypeInstanceToName((*logger.LoggingClient)(nil))
+
+// LoggingClientFrom helper function queries the DIC and returns the logger.loggingClient implementation.
+func LoggingClientFrom(get di.Get) logger.LoggingClient {
```
[the remark this answers is not in the document]
Comment (jinlinGuan) [reply pkg/bootstrap/container/logger.go]: renamed

File: pkg/bootstrap/handlers/message.go [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+// messages when the service first starts up successfully.
+func (h StartMessage) BootstrapHandler(
+	_ context.Context,
+	_ *sync.WaitGroup,
+	startupTimer startup.Timer,
+	dic *di.Container) bool {
+
+	logger := container.LoggerFrom(dic.Get)
+	logger.Info("Service dependencies resolved...")
+	logger.Info(fmt.Sprintf("Starting %s %s ", h.serviceKey, h.version))
```
Comment (cloudxxx8) [inline pkg/bootstrap/handlers/message.go:52]:
```suggestion
	logger.Infof("Starting %s %s ", h.serviceKey, h.version)
```

**Review by cloudxxx8 — CHANGES_REQUESTED** [review]

File: go.mod [hunk 13]
```diff
@@ -1,15 +1,17 @@
 module github.com/IOTechSystems/go-mod-edge-utils
 
-go 1.20
+go 1.21
 
 require (
 	github.com/go-kit/log v0.2.1
+	github.com/rcrowley/go-metrics v0.0.0-20201227073835-cf1acfcdf475
```
Comment (cloudxxx8) [inline go.mod:7]: we need to add metrics related code back to this repository in the next PR

**Review by cloudxxx8 — APPROVED** [review]
LGTM

**Pull request merged**